### ISSUE TITLE
Tests run in parallel despite NonParallelizable attribute

### ISSUE THREAD

**Issue by dign17** [nunit/nunit#3371]
I have the following definition for the assembly containing just a unit test:

    [assembly: Parallelizable(ParallelScope.Children)]

The code of the unit test looks like this:

    [TestFixture]
    public class SampleUnitTest
    {
        [Test]
        public void TestMethod1()
        {
            Thread.Sleep(10000);
        }

        [Test, TestCase(true), TestCase(false), NonParallelizable]
        public void TestMethod2(bool check)
        {
            if (check)
            {
                Thread.Sleep(5000);
            }
            else
            {
                Thread.Sleep(15000);
            }
        }
    }

I was wondering why the _TestMethod1()_ is executed in parallel to one of the test cases from _TestMethod2()_ even though the _NonParallelizable_ attribute is defined. If I understand the documentation correctly:

> This attribute is used to indicate that the test on which it appears may not be run in parallel with any other tests.
...
> When used on a test fixture or method, that test will be queued on the non-parallel queue and will not run while other tests marked as Parallelizable are being run.

each test should run sequentially. Please correct me if I'm missing something.

**Comment by dign17** [comment]
@CharliePoole Thank you for the detailed explanation, I will apply your strategy to solve my current issues.

If you say

> We specifically don't apply it to the test suite that is comprised of the two true/false test cases.

it sounds to me like a missing feature, isn't it?

**Comment by CharliePoole** [comment]
It's certainly an inconsistency, which could possibly be band-aided with a feature. It might have been better if we didn't create a suite for each parameterized method. Folks working on NUnit 4 may need to take note of it.

However, if we __did__ apply the attribute to the test suite rather than the cases, we would see other bugs - in fact, we have already seen them in an earlier release and what you have now is the fix.

**Comment by bormm** [comment]
Thanks for your response. I am working with @dign17 :-)

I understand that there is currently only one attribute to opt-out the Parallelizable attribute: the NonParallelizable attribute. This attribute is valued for the "child" tests that are created internally in case of specifying a testcase. These tests are then bundled in a new sub testsuite which is still flagged as parallelizable.

NonParallelizable is specified as "This attribute is used to indicate that the test on which it appears may **not be run in parallel with any other tests.** ... When used on a test fixture or method, **that test will be queued on the non-parallel queue** and will not run while other tests marked as Parallelizable are being run." I don't understand, even there are created some suites and tests internally, how these tests flagged as "NonParallelizable" can run parallel with any other test in any case. Shouldn't the handling on tests be identical however they are created? Or is the suite using his own queue?

**Comment by CharliePoole** [comment]
I answered here because I wrote the code a long time back. Someone on the team who is working on it currently (and who isn't on vacation) will have to take the discussion further. Hopefully, my comments will help them.

As I said, I personally consider the "invisible" test suite created for each method with test cases to be problematic. But it's something that probably cannot change since it has been this way for several years.

The wording of the docs is slightly imprecise since "test" can apply to either a test case or a test suite. When on a method, the attribute applies to the test cases. It has to be that way or you could not make cases run in parallel.

Even if some fix is made to this problem, you won't see it until a new release. So your best bet IMO is to mark the entire fixture (class) as non-parallelizable. Then, everything within the class will run on it's own, which is the safest option even if less efficient.

@nunit/framework-team It would be helpful if somebody else chimed in, since I'm doing this out of my head without trying it in code.

**Comment by blorger** [comment]
As far as I can tell problem exists only for parametrized tests marked as `NonParallelizable`.
Also, other tests will not start execution when non-parallelizable test is already executing. However parametrized non-parallelizable test will start execution regardless whether other test is already executing.

```C#
    using System.Diagnostics.CodeAnalysis;
    using System.Threading;
    using System.Threading.Tasks;

    using NUnit.Framework;

    [TestFixture]
    [Parallelizable(ParallelScope.All)]
    [SuppressMessage("Style", "CA1801:Remove unused parameter", Justification = "Testing test scheduling for parametrized tests.")]
#pragma warning disable IDE0060 // Remove unused parameter
    public static class ParallelTestExecutionControl
    {
        [Test]
        [Order(10)]
        [NonParallelizable]
        public static Task N1() => NonParallel();

        [Test]
        [Order(20)]
        [NonParallelizable]
        public static Task N2([Values(1, 2)] int x) => NonParallel();

        [Test]
        [Order(30)]
        public static Task P1() => Parallel();

        [Test]
        [Order(40)]
        public static Task P2([Values(1, 2)] int x) => Parallel();

        [Test]
        [Order(50)]
        public static Task P3() => Parallel();

        [Test]
        [Order(60)]
        public static Task P4([Values(1, 2)] int x) => Parallel();

        [Test]
        [Order(70)]
        public static Task P5() => Parallel();

        [Test]
        [Order(80)]
        [NonParallelizable]
        public static Task ZN1([Values(1, 2)] int x) => NonParallel();

        [Test]
        [Order(90)]
        [NonParallelizable]
        [TestCase(1)]
        [TestCase(2)]
        public static Task ZN2(int x) => NonParallel();

        [Test]
        [Order(100)]
        [NonParallelizable]
        public static Task ZN3() => NonParallel();


        private static int activeTestCount;

        private static async Task Parallel()
        {
            TestContext.WriteLine("Starting. Currently there are {0} active test instances.", Interlocked.Increment(ref activeTestCount));
            try
            {
                await Task.Delay(2000);
            }
            finally
            {
                TestContext.WriteLine("Finishing. Currently there are {0} active test instances.", Interlocked.Decrement(ref activeTestCount));
            }
        }

        private static async Task NonParallel()
        {
            int n = Interlocked.Increment(ref activeTestCount);
            try
            {
                Assert.That(n, Is.EqualTo(1), "Only single non-parallel test is executing at the same time.");
                await Task.Delay(300);
            }
            finally
            {
                int m = Interlocked.Decrement(ref activeTestCount);
                if (n == 1)
                {
                    Assert.That(m, Is.EqualTo(0), "Only single non-parallel test is finishing execution at the same time.");
                }
            }
        }
    }
```

**Comment by CharliePoole** [comment]
Cool! @blorger confirms experimentally my memory of how the code works. @rprouse Can you indicate whether this is viewed as a bug to be fixed in the current code or a design issue that needs to wait for a major release to be changed?

**Comment by heaths** [comment]
Can I assume the same problem is true of parameterized test fixtures, i.e. that each parameter variant - since it is a different test fixture - can run in parallel with other variants?

```csharp
using System;
using System.Threading.Tasks;
using NUnit.Framework;

[NonParallelizable]
[TestFixture(false)]
[TestFixture(true)]
public class Tests
{
    private static readonly Random rand = new Random();
    private readonly bool value;

    public Tests(bool value)
    {
        this.value = value;
    }

    [Test]
    public async Task Test1()
    {
        await Task.Delay(rand.Next(100, 5000));
        Console.WriteLine(value);
    }
}
```
I'm unable to repro the problem with this very simple test case and use it merely as an example, but we have some larger tests running against either REST endpoint or recorded sessions that occasionally fail with symptoms that indicate concurrency.

**Comment by CharliePoole** [comment]
@rprouse This is a couple of years old and there's still no indication of whether the NUnit project considers it a bug, for which somebody could contribute a fix, or if it's just the way it is for version 3.x. Also, is parallelization getting a re-design for 4.0?

**Comment by rprouse** [comment]
At this point there isn't a plan to redesign Parallelization in 4.0, so to me this is a bug that we would accept fixes for. It did take us several releases to get parallelism right though, so we'll want to be sure the fix doesn't cause regressions.

**Comment by CharliePoole** [comment]
@rprouse So just for clarity, the spec for a fix would be that `[NonParallelizable`] on a parameterized method means that __none__ of the individual test cases can run in parallel?

That would make sense to me. In fact, we do have docs that say that the effect of these attributes is inherited and we may even have had that behavior at some time in the past.

**Comment by madelson** [comment]
I'd love to see a fix for this. Tracking this down is not trivial.

**Comment by Bartleby2718** [comment]
I encountered this issue yesterday and would be interested in fixing this, but I don't have enough context and knowledge to do so. As far as I can tell, 
* `NonParallelizableAttribute` is the same as `ParallelScope.None`: https://github.com/nunit/nunit/blob/b34eba3ac1aa6957157857bddd116256c634afab/src/NUnitFramework/framework/Attributes/NonParallelizableAttribute.cs#L13-L18
* `ParallelScope` is an enum defined here: https://github.com/nunit/nunit/blob/b34eba3ac1aa6957157857bddd116256c634afab/src/NUnitFramework/framework/Attributes/ParallelScope.cs#L27-L32
* `ParallelExecutionStrategy.NonParallel` is chosen when `ParallelScope` is `Default` or `None`: https://github.com/nunit/nunit/blob/20a77efc520c2e0b397cec25a687dab7c8d22a5d/src/NUnitFramework/framework/Internal/Execution/ParallelWorkItemDispatcher.cs#L159-L167
* `ParallelWorkItemDispatcher.Start(WorkItem)` calls `Dispatch` with the right `ParallelExecutionStrategy`: https://github.com/nunit/nunit/blob/20a77efc520c2e0b397cec25a687dab7c8d22a5d/src/NUnitFramework/framework/Internal/Execution/ParallelWorkItemDispatcher.cs#L143-L157
* When non-parallel shifts assign a worker, queues are isolated only if `work.IsolateChildTests` is true: https://github.com/nunit/nunit/blob/20a77efc520c2e0b397cec25a687dab7c8d22a5d/src/NUnitFramework/framework/Internal/Execution/ParallelWorkItemDispatcher.cs#L53-L89
* `WorkItem.IsolateChildTests` defaults to false. https://github.com/nunit/nunit/blob/b34eba3ac1aa6957157857bddd116256c634afab/src/NUnitFramework/framework/Internal/Execution/WorkItem.cs#L151-L154
* A lock is used to isolate queues: https://github.com/nunit/nunit/blob/20a77efc520c2e0b397cec25a687dab7c8d22a5d/src/NUnitFramework/framework/Internal/Execution/ParallelWorkItemDispatcher.cs#L248-L264

However, I'm not sure why we have `if (work.IsolateChildTests)`. Shouldn't `IsolateQueues(work);` be called unconditionally in `OnStartNonParallelWorkItem`?

@CharliePoole Could you point me in the right direction? What should I read besides https://docs.nunit.org/articles/nunit/technical-notes/usage/Framework-Parallel-Test-Execution.html?

**Comment by CharliePoole** [comment]
To answer your question, we don't do it all the time because it's an expensive operation, which is only needed when we encounter a parallel test under a non-parallel fixture or suite. The first time we see such a method, we save the state of all the existing queues (think up to 100 or so) and create an entirely new set. If we didn't do this, parallel tests from other fixtures would be running in parallel with the non-parallel fixture. This was, in fact, a bug we fixed at one time. It's messy but works and I wouldn't risk changing it without a reason.

In fact, I don't see a reason. This issue doesn't have to do with the queues but with the fact that the test cases are defaulting to Parallel _within their containing suite_. That's how it's defined to work, so this is not a bug but a feature change, in my opinion. I don't think something is a bug because somebody prefers it to work a diffrent way. :-)

If I were doing this, I'd treat it as a feature and make sure I understood what was supposed to happen first. If nobody is doing that for you... post something that says "I intend to make it work like this...... Please let me know if you agree." You will have to include two things... What happens to the test cases with respect to parallelism? What happens to the suite containing them? Remember, the latter is a test too! I can think of three ways the [NonParallelizable] could work on a parameterized test method:

1. The containing suite is not parallelizable but the testcases remain parallelizable (current behavior).
2. The containing suite __and__ the individual cases are both parallelizable.
3. The containing suite is not affected (i.e. inherits its behavior) and the cases are parallelizable.

Ultimately, whatever the decision is, you only have to ensure that workitems are created to reflect it. Once that's done, I believe it will all work.

**Comment by madelson** [comment]
> This issue doesn't have to do with the queues but with the fact that the test cases are defaulting to Parallel within their containing suite. That's how it's defined to work, so this is not a bug but a feature change, in my opinion.

@CharliePoole can you explain what you mean by this? If I have a test project like so:

```csharp
[assembly: Parallelizable(ParallelScope.Fixtures)]

public class MyFixture
{
    [TestCase("a")]
    [TestCase("a")]
    [NonParallelizable]
    public void MyTest(string s) { ... }

    [Test]
    public void MyTest2() { ... }
}

public class MyFixture2
{
    [Test]
    public MyTest3() { ... } 
}
```

What are you saying that the behavior ought to be?

**Comment by CharliePoole** [comment]
@madelson I'm not saying what the behavior should be. If it's going to change, that's for somebody on the framework team to decide. What I'm saying is what the behavior is and has always been by design (option1) and what the possible alternate choices seem to be (options 2 and 3). Note that I was answering @Bartleby2718 who wants to work on it. They need an understanding of what the desired behavior is before spending time, I would think.

> Note edits to options 2 and 3, by the way.

To enlarge on what it is and has always done... 

* C# attributes apply to methods, classes, etc. not to other attributes. So, in your example, `NonParallelizable` applies to the `MyTest` method. 

* In NUnit, methods with the `Test` attribute represent a single test case, whereas methods with `TestCase` represent a suite of test cases.

* In the design of NUnit, the choice was made to have attributes on such a parameterized method apply to that suite, not to the individual test cases. So, in this case, `NonParallelizable` means that the suite itself is non-parallelizable, that is, it cannot run in parallel with any other suite or fixture. It says nothing about the test cases within the suite - they could be parallelizable among themselves or not.

I'm not saying that the above is "right" or "best" only that it's how it works, has always worked and is documented to work. It may very well have been a bad design choice in the first place. But changing it is actually a feature change, which has implications for how or even whether it should be changed in a 3.x release.

If we've learned anything about backward compatibility over the years, it's that _regretting_ a poor design choice doesn't mean you can just go ahead and change it to what you wish it had been without repercussions.

If you're asking me what I, as a user, think the behavior should be, I'd say option 2, because it's the safest. I _think_ it's what `[Parallelizable(ParallelScope.None)]` should do as well.

**Comment by madelson** [comment]
> So, in this case, NonParallelizable means that the suite itself is non-parallelizable, that is, it cannot run in parallel with any other suite or fixture. It says nothing about the test cases within the suite - they could be parallelizable among themselves or not.

How does that interact with the `ParallelScope.Fixtures` on the assembly level, though? Wouldn't that stop the individual test cases within the suite from running in parallel with each other or is that not how it stacks?

**Comment by CharliePoole** [comment]
There's no interaction at all. All fixtures are able to run in parallel, unless you make some exception on a particular fixture. The assembly-level setting is basically only a default.

Just FYI, as a coach, I generally discourage the use of `ParallelizableAttribute` at the assembly level. I believe it's better practice for every new fixture to start out life running non-parallel. Then, once it's working, you can test to see if it runs in parallel and only add the attribute to it in your CI once you are sure.

**Comment by madelson** [comment]
> There's no interaction at all. All fixtures are able to run in parallel, unless you make some exception on a particular fixture

@CharliePoole sorry if I'm just being dense here but I'm still a bit confused as to the specific behavior. My understanding is that the [example I gave above](https://github.com/nunit/nunit/issues/3371#issuecomment-1206788569) *without `[NonParallelizable]`* will run `MyFixture` in parallel with `MyFixture2`, but that within each fixture the tests will run in series. **Is that correct?** It is certainly my experience with NUnit.

If we then add in `[NonParallelizable]` on `MyTest`, should that change anything about the execution?

**Comment by CharliePoole** [comment]
1) Correct. 2) No it shouldn't. :-)

I should have emphasized that this issue has nothing whatsoever to do with fixtures. The structure of the example you gave is

* `MyFixture` (a class representing a TestFixture)
   * `MyTest` (a method representing a suite of test cases)
      * `MyTest("a")` (a test case)
      * `MyTest("a")` (a test case)
   * `MyTest2` (a method representing a single test case)
* `Fixture2` (a TestFixture)
   * `MyTest3` (a method representing a single test case)

There's no question about the two fixtures. They can run in parallel. The issue is what effect `[Parallelizable]` or `[NonParallelizable]` placed on `MyTest` should have. Does the attribute affect the suite represented by the method or each individual test case or both? Current behavior is that only the suite is affected.

In your specific example, it doesn't matter because both the suite and the test cases are already non-parallelizable, so the `NonParallelizable` attribute doesn't do anything anyway. However, the original example had everything set to parallel at the assembly level - `ParallelScope.Children`. In that case it matters significantly.

**Comment by bacar** [comment]
Apologies for dredging up an old issue.

On @madelson's example above, marking the parameterised `MyTest` `[NonParallelizable]` has no effect, as stated.
But marking the *non*-parameterised `MyTest2` does -- it makes it globally non-parallel, and stops it running concurrently with `MyTest3` in the other
fixture.

Setting aside any potentially breaking change to what `[NonParallelizable]` means on a parameterised method: is there *any* opt-in mechanism that makes the **cases** of `MyTest` globally non-parallel, short of changing the parallelism of the fixture or above? I couldn't find one.

If there isn't, is that a gap worth filling?

**Comment by OsirisTerje** [comment]
@manfred-brands @stevenaw  Comments ?

**Comment by stevenaw** [comment]
I think this is a neat idea. `NonParallelizable` is just an alias for `Parallelizable(ParallelScope.None)`. We also have `ParallelScope.Children` to have descendants opt in to parallelism, but no easy way to opt out.

I haven't checked what happens if we combine `Parallelizable(ParallelScope.Children | ParallelScope.None)`. If that _doesn't_ do what is being asked here then I wonder if we should change that. We could then also add a shorthand for `NonParallelizable(ParallelScope.Children)`

**Comment by bacar** [comment]
I did try out None|Children along with some other combinations of None|X - none of them seemed to prevent parallelism under Nunit 3/4/5

Related: #2003 discusses separation of parallel vs not from scope, and mentions a possible enum value `Cases`

**Comment by CharliePoole** [comment]
As originally designed, the attribute was intended to discourage people from setting tests to be run in parallel at a high level. We made it possible, even at the assembly level, but discouraged people from doing that in any guidance we gave. I understand that we may want to change that approach in the present day, but it may be useful to understand why we did it.

The `Parallelizable` attribute, obviously to us but not to all users, does not cause any test to actually be capable of running in parallel.  That's up to the user writing the code. The attribute merely notifies NUnit as to which tests are capable of running in parallel. It's entirely up to NUnit to decide how to actually schedule the tests. For example, there is no way to say that two arbitrary tests _must_ run in parallel or that two arbitrary tests _cannot_ run in parallel. Neither of those is impossible but they are not capabilities of the current implementation.

So what about `NonParallelizable`? Since it's just a synonym of `Parallelizable(None)` It is only capable of affecting the test on which it sits, in this case the method. The `ParallelScope` enum provides values we use to make child tests parallel but there is no way to force a child test to be non-parallel. In other words, by not encouraging making tests globally parallel, we didn't feel we needed to worry so much about how to undo that effect.

One way to fix this would be to give "NonParallelizable" an optional argument `bool includeDescendantTests=false`, which would cancel any parallelization for contained tests. Note that this should only be needed in an environment where `ParallelScope.All` has been set at a higher level. My rationale for using "DescendantTests" rather than "TestCases" is that this same problem could be said to exist for parameterized and generic fixtures as well.

A more general approach would be to allow a scope argument to `NonParallelizableAttribute` but my suspicion is that it would lead to more complications as well as confusion for users.

**Comment by manfred-brands** [comment]
> A more general approach would be to allow a scope argument to NonParallelizableAttribute but my suspicion is that it would lead to more complications as well as confusion for users.

I see it as a pair: `Parallelizable` and `NonParallelizable` where the closest one takes precedence.
If either is set on a `Test`, it overrides the one set on `TestFixtures`.
Both should have the same `Scope` options.

Examples below:

So:  `[assembly: [assembly: Parallelizable(ParallelScope.Fixtures)]` means all fixture can run in parallel, unless explicitly denied on a `TestFixture`:

```
[NonParallelizable(ParallelScope.Self)]
[TestFixture]
public class SomeFixtureRunOnItsOwn
{
}
```

This means that `SomeFixtureRunOnItsOwn` will not run unless none of the other fixtures is running.

```
[TestFixture]
public class FixtureWithParallellTests
{
    [TestCase(1)]
    [TestCase(2)]
    [Parallelizable(ParallelScope.Children)]
    public void TestsCanBeRunInParallel()
    {
    }
}
```

```
[TestFixture]
[Parallelizable(ParallelScope.Children)]
public class FixtureWithParallellTests
{
    [TestCase(1)]
    [TestCase(2)]
    [NonParallelizable(ParallelScope.Children)]
    public void TestsWillNotRunInParallel()
    {
    }

    [Test]
    public void ButCanRunInParallelWithThis()
    {
    }
}
```

**Comment by CharliePoole** [comment]
That could work. It's slightly more complicated for users but still uses the same scope options, except with a negative effect when associated with `NonParallelizableAttribute`, so it could work with the existing tree walk to set parallelism for each work item by just adding an extra parameter.

**Comment by bacar** [comment]
I'm struggling a little with:

> there is no way to say that ... two arbitrary tests cannot run in parallel

At the moment, marking any non-parameterized test method NonParallelizable seems (?) to have exactly the effect that they cannot run in parallel -- they do not seem to run in parallel with _any_ other test.

This seems consistent with _some_ readings of the docs:

- [`NonParallelizable`](https://docs.nunit.org/articles/nunit/writing-tests/attributes/nonparallelizable.html) _marks an assembly, fixture, or method so it is not run in parallel **with anything else**_
- _This Attribute is used to indicate that the test as well as its descendants may not be run in parallel with other tests_ -- [Parallel Test Execution](https://docs.nunit.org/articles/nunit/technical-notes/usage/Framework-Parallel-Test-Execution.html#nonparallelizableattribute) 

However the Parallel Test Execution page states that "Parallel class with non-parallel methods: The methods run sequentially, ... but may actually be running in parallel with other, unrelated methods from other classes". This seems at odds with the statements above, especially the first one, and I haven't been able to make this happen for a non-parameterised test. Obviously that's not proof that it cannot happen but I haven't managed to build a consistent understanding from multiple doc sources anyway. Can you give a concrete example where this actually could happen?

(For context: we have a lot of tests, they take nontrivial time to run in total, and we benefit significantly from parallelism and therefore have it set to Fixtures at the assembly level in our projects, having taken care to ensure the vast majority cannot conflict with one another; for a small number of tests we need to exercise an escape hatch that ensures the test runs in global isolation (typically where that test is checking the impact of changing some shared resource such as an environment variable, which may be read by many other tests). We generally strive to avoid leaning on that by  injecting values instead of leaning on a global resource -- but it's not always possible, nor the right tradeoff.)

**Comment by CharliePoole** [comment]
On Sat, Sep 5, 2026 at 2:31 AM bacar ***@***.***> wrote:

> *bacar* left a comment (nunit/nunit#3371)
> <https://github.com/nunit/nunit/issues/3371#issuecomment-5550898128>
>
> I'm struggling a little with:
>
> there is no way to say that ... two arbitrary tests cannot run in parallel
>
> The problem of two arbitrary tests is out of scope for this discussion but
since I brought it up, I'll try to explain briefly. :-)

Imagine you have 1000 tests, spread around various fixtures. Further, let
us say that they can all run in parallel but with one restriction: Test537
and Test942 may not run at the same time. Perhaps they use some resource in
common. That's what is meant by speaking of "arbitrary tests."

NUnit's implementation of parallelism is not designed to deal with this. Of
course, you could make both those tests non-parallel, but that would
prevent them from running in parallel with the other 998 tests, and you may
not want to do that. In that case your only option is to add some sort of
wait for that resource in the code of both tests. This will have some
effect on NUnit performance because you will occasionally tie up two of
NUnit's worker threads. Most likely it will be a small hit but that depends
on how many threads are available and how long-running the two tests are.

My only point in mentioning this was to say that a useful solution to your
problem has to work in the context of how NUnit's parallel test
implementation was created. Of course it's possible to completely redesign
and rewrite that implementation, but I doubt anyone wants to do that right
now. The current implementation has to work within the hierarchy of work
items executed by the dispatcher, which maps to the hierarchy of
tests. @Manfred
Brands ***@***.***> suggested solution does that.
